File: js/src/core/icon-grid-view.ts
import { BaseActionView } from './base/base-action-view';
import { ActionConfig } from '../types';
import { InvalidParameterError } from '../errors';

export type IconShape = 'circle' | 'square';

export interface IconGridContent {
    title?: string;
    intro?: string;
    shape?: IconShape;       // tile shape for all tiles, default 'circle'
    columns?: number;        // 2-6, default 4
    spacing?: number;        // px, default 12
    actions: ActionConfig[];
}

/**
 * Builds an SGUI `IconGrid` view — app-launcher style grid. Each action is a
 * square or circle image tile (from its `thumbnail`) with the `title` caption
 * underneath. Same action model + navigation as ActionGridView; only the
 * rendering differs (`desc` is not rendered).
 *
 * Builder methods: `setTitle`, `setIntro`, `setShape` ('circle'|'square'),
 * `setColumns` (2-6), `setSpacing`, `addIcon` (ergonomic tile: image + badge
 * first), plus the inherited action family (`addAction`, `removeAction`,
 * `updateAction`, `hasActions`), overridden here to keep the serialized
 * `content.actions` in sync.
 *
 * Extends {@link BaseActionView}. Created via the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized and signed into a v3 envelope
 * by serve().
 */
export class IconGridView extends BaseActionView {

    static fromJson(json: Record<string, unknown>): IconGridView {
        return IconGridView.fromJsonAs(IconGridView, 'IconGrid', json);
    }
    constructor(viewId: string, title: string, processId?: string) {
        super({
            id: viewId,
            type: 'IconGrid',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title: title,
            intro: '',
            shape: 'circle',
            columns: 4,
            spacing: 12,
            actions: this.actions,  // Reference to BaseActionView's actions
        } as IconGridContent;
    }

    /**
     * Sets the icon grid title
     */
    setTitle(title: string): this {
        (this.content as IconGridContent).title = title;
        return this;
    }

    /**
     * Sets the icon grid intro
     */
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    /**
     * Sets the tile shape (circle or square)
     */
    setShape(shape: IconShape): this {
        if (shape !== 'circle' && shape !== 'square') {
            throw new InvalidParameterError('shape', shape, "Shape must be 'circle' or 'square'");
        }
        (this.content as IconGridContent).shape = shape;
        return this;
    }

    /**
     * Sets the number of columns
     */
    setColumns(columns: number): this {
        if (columns < 2 || columns > 6) {
            throw new InvalidParameterError('columns', columns, 'Columns must be between 2 and 6');
        }
        (this.content as IconGridContent).columns = columns;
        return this;
    }

    /**
     * Sets the spacing between tiles
     */
    setSpacing(spacing: number): this {
        if (spacing < 0) {
            throw new InvalidParameterError('spacing', spacing, 'Spacing must be non-negative');
        }
        (this.content as IconGridContent).spacing = spacing;
        return this;
    }

    /**
     * Synchronize actions array to content (required by BaseActionView)
     */
    protected syncActionsToContent(): void {
        (this.content as IconGridContent).actions = this.actions;
    }

    /**
     * Overrides addAction (same order as the base class, to stay compatible
     * with addActions()) to keep content in sync.
     */
    override addAction(
        code: string,
        title: string,
        description?: string,
        thumbnail?: string,
        disabled: boolean = false,
        metadata?: Record<string, unknown>
    ): this {
        super.addAction(code, title, description, thumbnail, disabled, metadata);
        this.syncActionsToContent();
        return this;
    }

    /**
     * Adds an icon tile — ergonomic API for IconGrid: the image
     * (`thumbnail`) and `badge` first, since `desc` is not rendered.
     */
    addIcon(
        code: string,
        title: string,
        thumbnail?: string,
        badge?: string,
        disabled: boolean = false,
        metadata?: Record<string, unknown>
    ): this {
        super.addAction(code, title, undefined, thumbnail, disabled, metadata);
        const added = this.actions[this.actions.length - 1];
        if (added && badge !== undefined) {
            added.badge = badge;
        }
        this.syncActionsToContent();
        return this;
    }

    /**
     * Overrides removeAction to keep content in sync
     */
    override removeAction(actionCode: string): boolean {
        const result = super.removeAction(actionCode);
        if (result) {
            this.syncActionsToContent();
        }
        return result;
    }

    /**
     * Overrides updateAction to keep content in sync
     */
    override updateAction(actionCode: string, updates: Partial<ActionConfig>): boolean {
        const result = super.updateAction(actionCode, updates);
        if (result) {
            this.syncActionsToContent();
        }
        return result;
    }

    /**
     * Gets the tile shape
     */
    getShape(): IconShape {
        return (this.content as IconGridContent).shape || 'circle';
    }

    /**
     * Gets the number of columns
     */
    getColumns(): number {
        return (this.content as IconGridContent).columns || 4;
    }

    /**
     * Gets the spacing
     */
    getSpacing(): number {
        return (this.content as IconGridContent).spacing || 12;
    }

    /**
     * Checks whether the grid has any actions
     */
    hasActions(): boolean {
        return this.actions.length > 0;
    }
}

File: py/yeriasdk/views/icon_grid_view.py
"""
IconGridView - App-launcher style grid: square/circle image tiles with a
caption under each. Same action model + navigation as ActionGrid; only the
rendering differs. The mobile renderer shows image + title (+ optional badge);
`desc` is intentionally not rendered.
"""

from typing import Optional
from datetime import datetime

from .base_action_view import BaseActionView
from ..errors.exceptions import InvalidParameterError


class IconGridView(BaseActionView):
    """Builds an SGUI ``IconGrid`` view — app-launcher style grid. Each action
    is a square or circle image tile (from its ``thumbnail``) with the
    ``title`` caption underneath. Same action model + navigation as ActionGrid;
    only the rendering differs (``desc`` is not rendered).

    Builder methods: ``set_title``, ``set_intro``, ``set_shape``
    ('circle'|'square'), ``set_columns`` (2-6), ``set_spacing``, ``add_icon``
    (ergonomic tile: image + badge first), plus the inherited action family
    (``add_action``, ``remove_action``, ``update_action``, ``has_actions``),
    overridden here to keep the serialized ``content['actions']`` in sync.

    Extends ``BaseActionView``. Created via the YeriaApp/YeriaUI factory,
    populated with these builders, then serialized and signed into a v3
    envelope by ``serve()``.
    """
    @classmethod
    def from_json(cls, json_view):
        """Rehydrate a IconGrid view from a wire JSON payload."""
        return cls.from_json_as('IconGrid', json_view)


    def __init__(self, view_id: str, title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": view_id,
                "type": "IconGrid",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": title,
            "intro": "",
            "shape": "circle",
            "columns": 4,
            "spacing": 12,
            "actions": self._actions,  # Reference to actions from BaseActionView
        }

    def set_title(self, title: str) -> "IconGridView":
        """Set the icon grid title"""
        self.content["title"] = title
        return self

    def set_intro(self, intro: str) -> "IconGridView":
        """Set the introduction text displayed before the grid"""
        return self._set_intro_text("intro", intro)

    def set_shape(self, shape: str) -> "IconGridView":
        """Set the tile shape for all tiles ('circle' or 'square')"""
        if shape not in ("circle", "square"):
            raise InvalidParameterError(
                "shape", shape, "Shape must be 'circle' or 'square'"
            )
        self.content["shape"] = shape
        return self

    def set_columns(self, columns: int) -> "IconGridView":
        """Set the number of columns"""
        if columns < 2 or columns > 6:
            raise InvalidParameterError(
                "columns", columns, "Columns must be between 2 and 6"
            )
        self.content["columns"] = columns
        return self

    def set_spacing(self, spacing: int) -> "IconGridView":
        """Set spacing between tiles"""
        if spacing < 0:
            raise InvalidParameterError("spacing", spacing, "Spacing must be non-negative")
        self.content["spacing"] = spacing
        return self

    def _sync_actions_to_content(self) -> None:
        """Sync actions array to content (required by BaseActionView)"""
        self.content["actions"] = [
            {
                "code": action.code,
                "title": action.title,
                "desc": action.desc,
                "thumbnail": action.thumbnail,
                "badge": action.badge,
                "disabled": action.disabled if action.disabled else False,
                "metadata": action.metadata,
            }
            for action in self._actions
        ]

    def add_action(
        self,
        code: str,
        title: str,
        description: Optional[str] = None,
        thumbnail: Optional[str] = None,
        disabled: bool = False,
        metadata: Optional[dict] = None,
    ) -> "IconGridView":
        """Add an action (base order, compatible with add_actions) and sync"""
        super().add_action(code, title, description, thumbnail, disabled, metadata)
        self._sync_actions_to_content()
        return self

    def add_icon(
        self,
        code: str,
        title: str,
        thumbnail: Optional[str] = None,
        badge: Optional[str] = None,
        disabled: bool = False,
        metadata: Optional[dict] = None,
    ) -> "IconGridView":
        """Add an icon tile — ergonomic API (image + badge first; no desc)"""
        super().add_action(code, title, None, thumbnail, disabled, metadata)
        if badge is not None and self._actions:
            self._actions[-1].badge = badge
        self._sync_actions_to_content()
        return self

    def remove_action(self, action_code: str) -> bool:
        """Remove an action and sync to content"""
        result = super().remove_action(action_code)
        if result:
            self._sync_actions_to_content()
        return result

    def update_action(self, action_code: str, updates: dict) -> bool:
        """Update an action and sync to content"""
        result = super().update_action(action_code, updates)
        if result:
            self._sync_actions_to_content()
        return result

    def get_shape(self) -> str:
        """Get the tile shape"""
        return self.content.get("shape", "circle")

    def get_columns(self) -> int:
        """Get the number of columns"""
        return self.content.get("columns", 4)

    def get_spacing(self) -> int:
        """Get the spacing"""
        return self.content.get("spacing", 12)

    def has_actions(self) -> bool:
        """Check if the grid has actions"""
        return len(self._actions) > 0
